### PR TITLE
pipeline versioning control

### PR DESCRIPTION
**Pull request by diegok-unity**
Adding support for triggering specific pipeline versions and uploading new versions of an existing pipeline. Examples here: https://github.com/Unity-Technologies/ml-platform-demos/actions

### CODE REVIEW

File: kfp_utils.py [hunk 1]
```diff
@@ -53,7 +53,11 @@ def pipeline_compile(pipeline_function: object, v2_compatible: bool = False) ->
     return pipeline_name_zip
 
 
-def upload_pipeline(pipeline_name_zip: str, pipeline_name: str, kubeflow_url: str, client_id: str):
+def upload_pipeline(pipeline_name_zip: str,
+                    pipeline_name: str,
+                    kubeflow_url: str,
+                    client_id: str,
+                    pipeline_version_name: str = None):
```
Comment (vvnpan) [inline kfp_utils.py:61]: Can we add this to the doc strings arguments?

File: kfp_utils.py [hunk 2]
```diff
@@ -64,13 +68,22 @@ def upload_pipeline(pipeline_name_zip: str, pipeline_name: str, kubeflow_url: st
...
+        client.upload_pipeline(
+            pipeline_package_path=pipeline_name_zip,
+            pipeline_name=pipeline_name)
     return client
 
 
-def find_pipeline_id(pipeline_name: str, client: kfp.Client, page_size: str = 100, page_token: str = "") -> str:
+def find_pipeline_id(pipeline_name: str,
+                     client: kfp.Client,
+                     page_size: str = 100,
```
Comment (vvnpan) [inline kfp_utils.py:90]: this value is hard coded? Are there situations where there's more than 100 for teams running a lot of pipelines?
(same comment for all `page_size=100`

File: kfp_utils.py [hunk 3]
```diff
@@ -141,30 +193,42 @@ def read_pipeline_params(pipeline_parameters_path: str) -> dict:
...
+                 run_name: str = None):
     experiment_id = find_experiment_id(
         experiment_name=experiment_name, client=client, namespace=namespace)
-    if not experiment_id:
-        raise ValueError("Failed to find experiment with the name: {}".format(
-            experiment_name))
     logging.info(f"The expriment id is: {experiment_id}")
-    # [TODO] What would be a good way to name the jobs
-    job_name = pipeline_name + datetime.now().strftime("%Y_%m_%d_%H_%M_%S")
+    job_name = f"{pipeline_name}_{datetime.now().strftime('%Y_%m_%d_%H_%M_%S')}" if not run_name else run_name
```
Comment (vvnpan) [inline kfp_utils.py:237]: out of curiosity is this the same naming convention kubeflow uses in the UI?

Comment (diegok-unity) [reply kfp_utils.py:237]: Kubeflow convention seems to have changed, one example:

Run of demo-pipeline (28b56)

Looks like "Run of {pipeline version name} ({short hash})"

Do you think we should change this convention?

File: kfp_utils.py (hunk 1 above)
Comment (diegok-unity) [reply kfp_utils.py:61]: Docstrings updated across the file

File: kfp_utils.py (hunk 2 above)
Comment (diegok-unity) [reply kfp_utils.py:90]: This is just page size, the find functions iterate through all pages

File: kfp_utils.py (hunk 3 above)
Comment (vvnpan) [reply kfp_utils.py:237]: no I was just curious, this gives users better insight to when the job was launched

**Review by vvnpan — APPROVED** [review]

**Pull request merged**